### PR TITLE
[WIP] Improve CLI error messages for token-based auth (extensions)

### PR DESCRIPTION
**Pull request by molly-yu**
### WHY are these changes introduced?

Making errors more explicit and actionable (particularly when developers push their extensions either locally or on CI).
(https://github.com/Shopify/shopify-cli-planning/issues/69)

### WHAT is this pull request doing?

Case 1: On CI (no .env file) and missing arguments:
Screen Shot 2022-02-02 at 12 28 41 PM
Screen Shot 2022-02-02 at 12 28 54 PM
Screen Shot 2022-02-02 at 12 29 09 PM

Case 2: Local (.env file exists) and missing values in the file:
Screen Shot 2022-02-02 at 12 31 36 PM

Case 3: Local (no .env file):
Screen Shot 2022-02-02 at 12 31 36 PM

Note: 
- on CI, the error message shows when there is either no api-key or registration-id.
- locally, the error message only shows when there is no api-key (or no .env file).

### How to test your changes?

Run shopify extension push in the following scenarios:
- .env file exists but does not have certain values (e.g. api key,...)
- .env file does not exist (in the case of CI)

### Update checklist

- [ ] I've added a CHANGELOG entry for this PR (if the change is public-facing)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows).
- [x] I've left the version number as is (we'll handle incrementing this when releasing).
- [x] I've included any post-release steps in the section above.

### CODE REVIEW

**Comment by MeredithCastile** [comment]
@hannachen and @molly-yu and I chatted just now about the error messages. I wasn't able to reproduce the existing error messages as they came from Pedro. But we were able to make some superficial UI optimizations:

Current:
image

Tweaks:
image

The experience will change when we can print the variables in the CLI, as @amcaplan has proposed here: [shopify-cli/pull/1951](https://github.com/Shopify/shopify-cli/pull/1951) If possible, I'd like to print them in the error messages themselves. Like this:

image

I'll let the team and @stephszeto decide how to prioritize this according to effort/impact.

**Comment by stephszeto** [comment]
Err so the second screenshot / third is what we still need to implement? (I don't think the third is actually blocked on the `shopify app connect --print` command - we have that data already and can just implement the third option directly.) 

To me, the second option isn't that much better than the first - the third def is a marked improvement, but still feels like a later-stage optimization. Are we done covering the remaining error cases from [this issue](https://github.com/Shopify/shopify-cli-planning/issues/69
)? Finishing implementing those should be a higher priority than making the fixes above.

**Comment by MeredithCastile** [comment]
Yep, like I said, the middle screenshot is just superficial optimization to help me/us collectively come to greater understanding about error patterns.

Option 3 is what I'd like to launch with. Before we chatted yesterday, @molly-yu wasn't fully aware of @arielchernin's PR. So, unfortunately, she'll need to redo some of her work on those error messages. Good news: Molly and @hannachen have everything they need from me for now. So I'm going to reassign this issue to them on our team board.

**Comment by hannachen** [comment]
Sorry to jump into this convo late. My bad for not explaining this very well @MeredithCastile! I think I missed some details.

I think it makes sense for this iteration of improvements to include just this one:
> Tweaks: image

In order to print out the actual `api_key` and `uuid` values, they need to be cached somewhere like an `.env` file. However, when the CLI is used in a CI environment, we don't actually have the option to use an `.env` file, because it's generated by running the `connect` command.

CI environments are typically _not_ interactive, meaning we cannot run a `connect` command, then select an app/script to push to. The only way the CLI can get this information in a CI environment is if it's passed using flags. I think that's the reason why @amcaplan mentioned that particular case would never happen.

File: lib/project_types/extension/loaders/project.rb
Comment (hannachen) [reply lib/project_types/extension/loaders/project.rb:36]: I think having different error message instructions based on `tty?` makes sense, as the instructions might be different. i.e. we can ask the developer to run a `connect` command to generate the `.env` file, but on CI, passing them in as flags is the only option.

**Review by hannachen — APPROVED** [review]
I feel that this is a good improvement compared to what we had previously :) Additional details can be iterated on top of this!

**Pull request merged**
